### PR TITLE
add asprintf

### PR DESCRIPTION
**Pull request by xoviat**
reasons why it's done this way:

1. I prefer to keep header files free of code.
2. windows functionality should not interfere with linux readability. windows should try to "also" work in addition to linux; therefore linux functions should be created on windows as a wrapper.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
Please add license headers (new LLVM) to new files. Check new line at end of file.

**Comment by xoviat** [comment]
I will add the license after I get an approval from @kaadam.

**Comment by kiranchandramohan** [comment]
Is the approval from @kaadam because he is the author? If so it might be good to add his as a co-author.

**Comment by xoviat** [comment]
No, he is not an author. I am trying to get him to approve my pull requests, because that's what I said I would do earlier. If there's no approval in a few days, then I will proceed without him.

File: tools/flang2/flang2exe/asprintf.c
Comment (kaadam) [inline tools/flang2/flang2exe/asprintf.c:1]: Please add license header here too.

File: tools/flang2/flang2exe/asprintf.c
Comment (kaadam) [inline tools/flang2/flang2exe/asprintf.c:34]: Please add a new line.

**Review by kaadam — COMMENTED** [review]
It looks good to me all in all. Thanks for the patch. Just a nit: please update your change based on the comments.

**Comment by kaadam** [comment]
Just on more thing: Can you add a bit more info to the commit message as you left a description in asprintf.c. Thanks.

File: .gitignore
Comment (kaadam) [inline .gitignore:4]: Nit: please add a new line here

**Review by kaadam — APPROVED** [review]

File: tools/flang2/flang2exe/asprintf.c
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/asprintf.c:33]: Is this function used anywhere in the codebase?

File: tools/flang2/flang2exe/CMakeLists.txt
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/CMakeLists.txt:25]: Files in flang2 were all converted to cpp files sometime back. Is it possible to have this file also as a cpp file?

File: tools/flang2/flang2exe/asprintf.c
Comment (bryanpkc) [reply tools/flang2/flang2exe/asprintf.c:33]: It is only used in kmpcutil.cpp. IMO we should rewrite the call site to use a different function if possible, rather than add the wrapper and additional maintenance.

**Comment by xoviat** [comment]
@bryanpkc The alternative is this: https://github.com/kaadam/flang/blob/windows/tools/flang2/flang2exe/kmpcutil.cpp#L530

Do you like that approach better?

**Comment by kiranchandramohan** [comment]
@bryanpkc could you reply?

**Comment by bryanpkc** [comment]
> @bryanpkc The alternative is this: https://github.com/kaadam/flang/blob/windows/tools/flang2/flang2exe/kmpcutil.cpp#L530

I am fine either way; personally I wouldn't create a new file just for one call site, but since you have done the work, it is harmless to commit the change and make `vasprintf` available for potential  future call sites.

**Comment by bryanpkc** [comment]
The return value of `vasprintf` is not checked in kmpcutil.cpp, and `nm` may end up being undefined. Should we check for this case and throw an ICE if it happens?

File: tools/flang2/flang2exe/kmpcutil.cpp
Comment (xoviat) [inline tools/flang2/flang2exe/kmpcutil.cpp:534]: @bryanpkc Is this fine?

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

File: tools/flang2/flang2exe/kmpcutil.cpp
Comment (bryanpkc) [reply tools/flang2/flang2exe/kmpcutil.cpp:534]: LGTM, thanks.

**Review by bryanpkc — APPROVED** [review]
LGTM.

**Comment by kiranchandramohan** [comment]
@xoviat can you squash the commits?

**Review by shivaramaarao — APPROVED** [review]

**Comment by kiranchandramohan** [comment]
ping @xoviat. Should the commits be squashed or is it oK to merge?

**Comment by xoviat** [comment]
@kiranchandramohan done.

**Pull request merged**
